### PR TITLE
4.x : Workspace Refactoring - Add a property to a model

### PR DESCRIPTION
**Pull request by deepakrkris**
Feature to add a model property to a workspace graph
- [X] - add a model-property loopback model
- [X] - add method to connector to handle createModelProperty
- [X] - add a method to workspace tasks for addModelProperty
- [X] - create a new class for model property in workspace graph
- [X] - add a new gherkin cucumber test case

### CODE REVIEW

File: test/acceptance-tests/workspace-devtime/step-definitions/create-models.js [hunk 1]
```diff
@@ -39,7 +40,38 @@ module.exports = function() {
...
-    expect(testsuite.expectedModel).to.eql(storedModel);
+    expect(testsuite.expectedModel).to.eql(storedModel._content);
+    next();
+  });
+
+  this.Given(/^the model '(.+)' exists$/, function(modelName, next) {
+    testsuite.modelId = 'common.' + modelName;
+    var workspace = workspaceManager.getWorkspace();
+    var storedModel = workspace.getModel(testsuite.modelId);
+    expect(storedModel).to.not.to.be.undefined;
```
Comment (superkhau) [inline test/acceptance-tests/workspace-devtime/step-definitions/create-models.js]: undefined()

Comment (superkhau) [inline test/acceptance-tests/workspace-devtime/step-definitions/create-models.js]: and fix the sentence

```
expect(storedModel).to.not.be.undefined(); // be.an('object')? and assert some properties on this object? instanceOf?

File: test/acceptance-tests/workspace-devtime/step-definitions/create-models.js [hunk 2]
```diff
@@ -39,7 +40,38 @@ module.exports = function() {
...
+    expect(storedModel).to.not.to.be.null;
```
Comment (superkhau) [inline test/acceptance-tests/workspace-devtime/step-definitions/create-models.js]: and fix the sentence

```
expect(storedModel).to.not.be.null(); // be.an('object')? and assert some properties on this object? instanceOf?

**Comment by superkhau** [comment]
@deepakrkris BTW, do you plan to use `const` + `let` instead of `var` since you're using ES6 in some places already?

File: test/acceptance-tests/workspace-devtime/step-definitions/create-models.js (hunk 1 above)
Comment (deepakrkris) [inline test/acceptance-tests/workspace-devtime/step-definitions/create-models.js]: agreed

File: test/acceptance-tests/workspace-devtime/step-definitions/create-models.js (hunk 2 above)
Comment (deepakrkris) [inline test/acceptance-tests/workspace-devtime/step-definitions/create-models.js]: agreed

File: common/models/data-source-definition.js [hunk 3]
```diff
@@ -16,8 +16,8 @@ module.exports = function(DataSourceDefinition) {
         cb = options;
         options = null;
       }
-      var connector = DataSourceDefinition.getConnector();
-      var id = data.id;
+      let connector = DataSourceDefinition.getConnector();
```
Comment (superkhau) [inline common/models/data-source-definition.js]: const

Comment (deepakrkris) [inline common/models/data-source-definition.js]: agreed

Comment (deepakrkris) [inline common/models/data-source-definition.js]: @superkhau , done the changes

File: .eslintrc [hunk 4]
```diff
@@ -1,3 +1,6 @@
 {
-  "extends": "loopback"
+  "extends": "loopback",
+  "rules": {
+    "no-unused-expressions": "off"
+  }
```
Comment (superkhau) [inline .eslintrc]: Do not add no-unused-expresssions rule.

File: common/models/model-property.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        options = null;
+      }
+      const connector = ModelProperty.getConnector();
+      let id = data.id;
+      let tokens = id.split('.');
+      if (tokens && tokens.length === 3) {
+        let facet = tokens[0];
+        let modelName = tokens[1];
+        let propertyName = tokens[2];
+        let modelId = facet + '.' + modelName;
```
Comment (superkhau) [inline common/models/model-property.js]: all the above are consts since they are all read only

File: component/datamodel/graph/index.js [hunk 6]
```diff
@@ -13,8 +13,8 @@ class Graph {
     this._cache[name] = {};
   }
   addNode(node) {
-    var domain = node._domain;
-    var name = node._name;
+    let domain = node._domain;
+    let name = node._name;
```
Comment (superkhau) [inline component/datamodel/graph/index.js]: const

File: component/tasks.js [hunk 7]
```diff
@@ -10,17 +11,26 @@ var DataSource = require('./datamodel/datasource');
  */
 class Tasks {
   addModel(modelId, modelDef, cb) {
-    var workspace = this;
+    let workspace = this;
```
Comment (superkhau) [inline component/tasks.js]: const

File: component/tasks.js [hunk 8]
```diff
@@ -10,17 +11,26 @@ var DataSource = require('./datamodel/datasource');
...
     //Model is a self-aware node which adds itself to the Workspace graph
     new Model(workspace, modelId, modelDef);
     cb(null, modelDef);
   }
   addDataSource(id, datasource, cb) {
-    var workspace = this;
+    let workspace = this;
```
Comment (superkhau) [inline component/tasks.js]: const

File: component/workspace.js [hunk 9]
```diff
@@ -17,20 +17,24 @@ class Workspace extends Graph {
     mixin(this, Tasks.prototype);
   }
   getModel(modelId) {
-    var model = this.getNode('ModelDefinition', modelId);
-    return model._content;
+    let model = this.getNode('ModelDefinition', modelId);
```
Comment (superkhau) [inline component/workspace.js]: const

File: component/workspace.js [hunk 10]
```diff
@@ -17,20 +17,24 @@ class Workspace extends Graph {
...
+    return model;
   }
   getDataSource(id) {
-    var ds = this.getNode('DataSource', id);
-    return ds._content;
+    let ds = this.getNode('DataSource', id);
+    return ds;
```
Comment (superkhau) [inline component/workspace.js]: `return this.getNode('DataSource', id);`

or

`const ds...`

File: component/workspace.js [hunk 11]
```diff
@@ -17,20 +17,24 @@ class Workspace extends Graph {
...
+  }
+  getModelProperty(id) {
+    let property = this.getNode('ModelProperty', id);
+    return property;
   }
 };
 
 function mixin(target, source) {
-  var attributes = Object.getOwnPropertyNames(source);
+  let attributes = Object.getOwnPropertyNames(source);
```
Comment (superkhau) [inline component/workspace.js]: const

File: component/workspace.js [hunk 12]
```diff
@@ -17,20 +17,24 @@ class Workspace extends Graph {
...
   attributes.forEach(function(ix) {
     if (typeof source[ix] === 'function') {
-      var mx = source[ix];
+      let mx = source[ix];
```
Comment (superkhau) [inline component/workspace.js]: const

File: connector/connector.js [hunk 13]
```diff
@@ -1,19 +1,24 @@
 'use strict';
-var app = require('../server/server.js');
-var connector = app.dataSources.db.connector;
-var WorkspaceManager = require('../component/workspace-manager.js');
+let app = require('../server/server.js');
+let connector = app.dataSources.db.connector;
```
Comment (superkhau) [inline connector/connector.js]: above 2 are consts

File: test/acceptance-tests/workspace-devtime/step-definitions/create-datasource.js [hunk 14]
```diff
@@ -1,17 +1,17 @@
...
-var loopback = require('loopback');
-var path = require('path');
-var util = require('util');
-var workspaceManager = require('../../../../component/workspace-manager.js');
-var DataSourceDefinition = app.models.DataSourceDefinition;
+let app = require('../../../../');
+let expect = require('../../../helpers/expect');
+let loopback = require('loopback');
+let path = require('path');
+let util = require('util');
```
Comment (superkhau) [inline test/acceptance-tests/workspace-devtime/step-definitions/create-datasource.js]: above are all consts

File: test/acceptance-tests/workspace-devtime/step-definitions/create-datasource.js [hunk 15]
```diff
@@ -1,17 +1,17 @@
...
+const workspaceManager = require('../../../../component/workspace-manager.js');
+const DataSourceDefinition = app.models.DataSourceDefinition;
 app.on('booted', function() {
   app.emit('ready');
 });
 
 module.exports = function() {
-  var testsuite = this;
+  let testsuite = this;
```
Comment (superkhau) [inline test/acceptance-tests/workspace-devtime/step-definitions/create-datasource.js]: const

**Pull request merged**